src/kaleidoscope_internal/event_dispatch.h

```diff
@@ -0,0 +1,141 @@
+                                                                          __NL__ \
+   } 	                                                                  __NL__ \
+                                                                          __NL__ \
+   namespace kaleidoscope {                                               __NL__ \
+                                                                          __NL__ \
+     EventHandlerResult Hooks::HOOK_NAME SIGNATURE {                      __NL__ \
+        return kaleidoscope_internal::EventDispatcher::template           __NL__ \
+        apply<kaleidoscope_internal::EventHandler_ ## HOOK_NAME>          __NL__ \
+             (__VA_ARGS__);                                               __NL__ \
+      }                                                                   __NL__ \
+                                                                          __NL__ \
+   }
+
+// _KALEIDOSCOPE_INIT_PLUGINS builds the loops that execute the plugins'
+// implementations of the various event handlers.
+//
+// Its arguments are a list of references to plugin instances that have been
+// instantiated in the global scope.
+
+// EventDispatcher::apply() implements a compile time for-each loop over all
+// plugins. The compiler automatically optimizes away calls to any plugin that
+// doesn't implement an EventHandler for a given hook.
+
+#define _KALEIDOSCOPE_INIT_PLUGINS(...)                                       __NL__ \
+  namespace kaleidoscope_internal {                                           __NL__ \
+  struct EventDispatcher {                                                    __NL__ \
+                                                                              __NL__ \
+    /* Iterate through plugins, calling each one's event handler with         __NL__ \
+       the arguments passed to the hook  */                                   __NL__ \
+    template<typename EventHandler__, typename... Args__ >                    __NL__ \
+    static kaleidoscope::EventHandlerResult apply(Args__&&... hook_args) {    __NL__ \
+                                                                              __NL__ \
+      kaleidoscope::EventHandlerResult result;                                __NL__ \
+      MAP(_INLINE_EVENT_HANDLER_FOR_PLUGIN, __VA_ARGS__)                      __NL__ \
+                                                                              __NL__ \
+      return result;                                                          __NL__ \
+    }                                                                         __NL__ \
+  };                                                                          __NL__ \
+                                                                              __NL__ \
+  }                                                                           __NL__ \
+  _REGISTER_EVENT_HANDLER(onSetup,false,())                                   __NL__ \
```

Comment:
Proposal: Lets use extended [X Macros](https://en.wikipedia.org/wiki/X_Macro) here for greater flexibility with processing hooks in other places (or in the future).

```cpp
// Variadic arguments of _FOR_EACH_EVENT_HANDLER are added to the parameter list of OP.
// This may optionally be used to pass additional information to the operation OP.
//
#define _FOR_EACH_EVENT_HANDLER(OP, ...)                                __NL__ \
   OP(onSetup, false,                                                   __NL__ \
               (),(), ##__VA_ARGS__)                                    __NL__ \
   OP(beforeEachCycle, false,                                           __NL__ \
               (), (), ##__VA_ARGS__)                                   __NL__ \
   OP(onKeyswitchEvent, true,                                           __NL__ \
               (Key &mappedKey, byte row, byte col, uint8_t keyState),  __NL__ \
               (mappedKey, row, col, keyState),                         __NL__ \
                ##__VA_ARGS__)                                          __NL__ \
   OP(beforeReportingState, false,                                      __NL__ \
               (),(),##__VA_ARGS__)                                     __NL__ \
   OP(afterEachCycle, false,                                            __NL__ \
               (),(),##__VA_ARGS__)      

// Invokes _FOR_EACH_EVENT_HANDLER and uses _REGISTER_EVENT_HANDLER for OP
//
_FOR_EACH_EVENT_HANDLER(_REGISTER_EVENT_HANDLER)
```

Note, the `##__VA_ARGS__` ([explanation](https://gcc.gnu.org/onlinedocs/gcc-4.8.2/gcc/Variadic-Macros.html#Variadic-Macros)) which is necessary to be able to pass zero or more additional arguments to `_FOR_EACH_EVENT_HANDLER` that are passed through to the operation `OP`.

Future applications could e.g. be something like this:
```cpp
// This macro has the same signature as _REGISTER_EVENT_HANDLER, except for the additional 
// argument PLUGIN_CLASS. This is generated from an additional variadic argument of 
// _FOR_EACH_EVENT_HANDLER that is passed on to OP where it ends up as PLUGIN_CLASS.
//
#define DEBUG_EVENT_HANDLERS_OF_PLUGIN(HOOK_NAME, SHOULD_ABORT_ON_CONSUMED_EVENT, SIGNATURE, ARGS_LIST, PLUGIN_CLASS) \
   /* Do something special with class PLUGIN_CLASS and its handlers */ \
   ...

_FOR_EACH_EVENT_HANDLER(DEBUG_EVENT_HANDLERS_OF_PLUGIN, kaleidoscope::MyFancyPluginClass)
```

Reply:
I think this is roughly what both @obra and me wanted to end up with eventually, but went to sleep before the last mile.

Thanks! :)

Reply:
As it is way easier for me to write code than to write explanations, I opened a PR #318 that covers the proposed changes.